File: desktop/src/utils/ebayMarks.js
// desktop/src/utils/ebayMarks.js
// Spec H3b §4.4/§5.4 -- eBay-Marken und Einrichtungs-Check-Liste für die Anzeige. ZWILLING:
// android/app/src/main/java/com/example/yugiohscanner/ml/EbayMarks.kt. Gemeinsame Fixture: docs/fixtures/ebay/marks.json.
// Wer eine Fassung ändert, ändert beide. status: undefined = lädt, null = kein Stand (Tabellen fehlen/nie verbunden).

export const MAX_OWN_PHOTOS = 12;
const WAITING = { kind: 'wartet', text: 'wartet auf eBay', url: null, retry: false };

const isActive = (l) => l.status === 'aktiv' && !l.deleted;
const webUrl = (u) => (typeof u === 'string' && /^https:\/\//i.test(u.trim()) ? u.trim() : null);

export function ebayMark(listing, row, status) {
  if (!listing || listing.channel_id !== 'ebay') return null;
  if (status === undefined) return { kind: 'laden', text: '…', url: null, retry: false };
  const active = isActive(listing);
  if (!status || !row) return active ? WAITING : null;
  if (row.environment !== status.environment) {
    return active ? WAITING : { kind: 'andere', text: 'auf eBay beendet (andere Umgebung)', url: null, retry: false };
  }
  if (row.state === 'online') return { kind: 'online', text: 'auf eBay online', url: webUrl(row.item_url), retry: false };
  if (row.state === 'fehler') {
    return { kind: 'fehler', text: `eBay-Fehler: ${row.error || 'unbekannt'}`, url: webUrl(row.item_url), retry: active };
  }
  if (row.state === 'beendet') return { kind: 'beendet', text: 'auf eBay beendet', url: null, retry: false };
  return active ? WAITING : null;
}

export function setupItems(status) {
  const s = status || {};
  const named = (label, ok, name) => ({ label: ok && name ? `${label}: ${name}` : label, ok: !!ok });
  return [
    { label: 'Mit eBay verbunden', ok: !!s.connected },
    named('Zahlungsrichtlinie', s.has_payment_policy, s.payment_policy_name),
    named('Versandrichtlinie', s.has_fulfillment_policy, s.fulfillment_policy_name),
    named('Rücknahmerichtlinie', s.has_return_policy, s.return_policy_name),
    named('Artikelstandort', s.has_location, s.location_key),
  ];
}
export const setupOk = (status) => setupItems(status).every((i) => i.ok);

// Spec §4.3: 30 Tage vor Ablauf des Refresh-Tokens ein Hinweis. today = 'YYYY-MM-DD' (lokal).
export function expiryText(status, today) {
  if (!status || !status.connected || !status.refresh_expires_at) return null;
  const day = String(status.refresh_expires_at).slice(0, 10);
  const utc = (s) => Date.UTC(Number(s.slice(0, 4)), Number(s.slice(5, 7)) - 1, Number(s.slice(8, 10)));
  const days = Math.round((utc(day) - utc(today)) / 86400000);
  if (days > 30) return null;
  return `Verbindung läuft am ${day.slice(8, 10)}.${day.slice(5, 7)}.${day.slice(0, 4)} ab – bitte neu verbinden.`;
}

export const photoCountText = (n) => `${n}/${MAX_OWN_PHOTOS} Fotos`;

File: android/app/src/main/java/com/example/yugiohscanner/ml/EbayMarks.kt
package com.example.yugiohscanner.ml

import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * Spec H3b §4.4/§5.4 -- eBay-Marken und Einrichtungs-Check-Liste. ZWILLING von desktop/src/utils/ebayMarks.js,
 * gemeinsame Fixture docs/fixtures/ebay/marks.json. Wer eine Fassung ändert, ändert beide.
 * Namensabweichungen: JS-status `undefined` = [StatusState.Loading], `null` = [StatusState.None]; Felder camelCase.
 */
object EbayMarks {
    const val MAX_OWN_PHOTOS = 12

    data class Mark(val kind: String, val text: String, val url: String?, val retry: Boolean)
    data class ListingHead(val channelId: String, val status: String, val deleted: Boolean)
    data class Row(val environment: String, val state: String, val itemUrl: String?, val error: String?)
    data class Status(
        val environment: String, val connected: Boolean, val refreshExpiresAt: String?,
        val hasPaymentPolicy: Boolean, val paymentPolicyName: String?,
        val hasFulfillmentPolicy: Boolean, val fulfillmentPolicyName: String?,
        val hasReturnPolicy: Boolean, val returnPolicyName: String?,
        val hasLocation: Boolean, val locationKey: String?,
    )
    sealed interface StatusState {
        data object Loading : StatusState
        data object None : StatusState
        data class Known(val status: Status) : StatusState
    }
    data class SetupItem(val label: String, val ok: Boolean)

    private val WAITING = Mark("wartet", "wartet auf eBay", null, false)
    private val HTTPS = Regex("^https://", RegexOption.IGNORE_CASE)

    fun webUrl(u: String?): String? = u?.trim()?.takeIf { HTTPS.containsMatchIn(it) }

    fun mark(l: ListingHead, row: Row?, st: StatusState): Mark? {
        if (l.channelId != "ebay") return null
        if (st is StatusState.Loading) return Mark("laden", "…", null, false)
        val active = l.status == "aktiv" && !l.deleted
        val s = (st as? StatusState.Known)?.status
        if (s == null || row == null) return if (active) WAITING else null
        if (row.environment != s.environment) {
            return if (active) WAITING else Mark("andere", "auf eBay beendet (andere Umgebung)", null, false)
        }
        return when (row.state) {
            "online" -> Mark("online", "auf eBay online", webUrl(row.itemUrl), false)
            "fehler" -> Mark("fehler", "eBay-Fehler: ${row.error?.takeIf { it.isNotEmpty() } ?: "unbekannt"}", webUrl(row.itemUrl), active)
            "beendet" -> Mark("beendet", "auf eBay beendet", null, false)
            else -> if (active) WAITING else null
        }
    }

    fun setupItems(s: Status?): List<SetupItem> {
        fun named(label: String, ok: Boolean, name: String?) =
            SetupItem(if (ok && !name.isNullOrEmpty()) "$label: $name" else label, ok)
        return listOf(
            SetupItem("Mit eBay verbunden", s?.connected == true),
            named("Zahlungsrichtlinie", s?.hasPaymentPolicy == true, s?.paymentPolicyName),
            named("Versandrichtlinie", s?.hasFulfillmentPolicy == true, s?.fulfillmentPolicyName),
            named("Rücknahmerichtlinie", s?.hasReturnPolicy == true, s?.returnPolicyName),
            named("Artikelstandort", s?.hasLocation == true, s?.locationKey),
        )
    }
    fun setupOk(s: Status?): Boolean = setupItems(s).all { it.ok }

    /** Spec §4.3: 30 Tage vor Ablauf ein Hinweis. [today] = "YYYY-MM-DD" (lokal). */
    fun expiryText(s: Status?, today: String): String? {
        if (s == null || !s.connected || s.refreshExpiresAt.isNullOrEmpty()) return null
        val day = s.refreshExpiresAt.take(10)
        val days = ChronoUnit.DAYS.between(LocalDate.parse(today), LocalDate.parse(day))
        if (days > 30) return null
        return "Verbindung läuft am ${day.substring(8, 10)}.${day.substring(5, 7)}.${day.substring(0, 4)} ab – bitte neu verbinden."
    }

    fun photoCountText(n: Int): String = "$n/$MAX_OWN_PHOTOS Fotos"
}
